File: domain/FullMove.ts
/*
* ===============================================================================
* File Name          : FullMove.ts
* Creation Date      : 2025-11-01
* Last Modified      : 
* Version            : 1.0.0
* License            : 
* Author             : Rigoberto L. Salgado Reyes
* Contact            : rlsalgado2006@gmail.com
* ===============================================================================
*/

import { Move } from "(src)/domain/Move";
import { PieceKind } from "(src)/domain/types";

class FullMove {
	private readonly _moves: Move[];
	private readonly _score: number;
	private readonly names: Record<number, string> = {1: "BLACK", 2: "WHITE", 3: "NEUTRON"};

	constructor(moves: Move[], score: number) {
		if (moves.length !== 4) throw new Error("A FullMove must contain exactly four moves");
		this._moves = moves;
		this._score = score;
	}

	get moves(): Move[] {
		return this._moves;
	}

	get score(): number {
		return this._score;
	}

	// noinspection JSUnusedGlobalSymbols
	clone() {
		const clonedMoves = this.moves.map(move => move.clone());
		return new FullMove(clonedMoves, this.score);
	}

	private kind2Name(pieceKind: PieceKind) {
		return this.names[pieceKind] || "NO KIND";
	}

	toString() {
		if (this.moves.length) {
			const piece1Kind = this.kind2Name(this.moves[1].kind);
			const piece2Kind = this.kind2Name(this.moves[3].kind);
			return `${piece1Kind}: ${this.moves[0].toString()}-${this.moves[1].toString()}, ` +
				`${piece2Kind}: ${this.moves[2].toString()}-${this.moves[3].toString()}`;
		}

		return `EMPTY FULLMOVE with score = ${this.score}`;
	}

	// noinspection JSUnusedGlobalSymbols
	toJSON() {
		return {
			__typename: "FullMove",
			moves: this.moves.map(move => move.toJSON()),
			score: this.score
		};
	}

	static fromJSON(json: any): FullMove {
		const moves = json.moves.map((moveJson: any) => Move.fromJSON(moveJson));
		return new FullMove(moves, json.score);
	}

	empty(): boolean {
		return !this.moves.length;
	}
}

export { FullMove };

File: native/src/FullMove.cpp
/**
 * Authors:
 * Rigoberto Leander Salgado Reyes <rlsalgado2006@gmail.com>
 *
 * Copyright 2018 by Rigoberto Leander Salgado Reyes.
 *
 * This program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 * Modified 2025 by Rigoberto Leander Salgado Reyes.
 */

#include <FullMove.h>

FullMove::FullMove(const std::initializer_list<Move> moves, const int score) {
    this->moves.reserve(moves.size());
    for (auto &m : moves) this->moves.emplace_back(std::make_unique<Move>(m));
    this->score = score;
}

FullMove::FullMove(std::vector<std::unique_ptr<Move>> &&mv, const int s) : moves(std::move(mv)), score(s) {
}

FullMove::FullMove(const FullMove &other) : score(other.score) {
    moves.reserve(other.moves.size());
    for (const auto &m : other.moves) {
        moves.emplace_back(std::make_unique<Move>(*m));
    }
}

FullMove &FullMove::operator=(const FullMove &other) {
    if (this != &other) {
        std::vector<std::unique_ptr<Move>> tmp;
        tmp.reserve(other.moves.size());
        for (const auto &m : other.moves) {
            tmp.emplace_back(std::make_unique<Move>(*m));
        }
        moves = std::move(tmp);
        score = other.score;
    }
    return *this;
}

std::unique_ptr<FullMove> FullMove::clone() const {
    return std::make_unique<FullMove>(*this);
}

std::string FullMove::kind2Name(PieceKind &pieceKind) const {
    switch (pieceKind) {
        case PieceKind::BLACK:
            return "BLACK";
        case PieceKind::WHITE:
            return "WHITE";
        case PieceKind::NEUTRON:
            return "NEUTRON";
        default:
            return "NO KIND";
    }
}

bool FullMove::empty() const {
    return moves.empty();
}

// std::ostream &operator<<(std::ostream &ostr, const FullMove &fullMove) {
//     if (!fullMove.empty()) {
//         auto piece1Kind = fullMove.kind2Name(fullMove.moves->at(1)->kind);
//         auto piece2Kind = fullMove.kind2Name(fullMove.moves->at(3)->kind);
//         ostr << piece1Kind << ": " << *fullMove.moves->at(0) << "-" << *fullMove.moves->at(1) << " ";
//         ostr << piece2Kind << ": " << *fullMove.moves->at(2) << "-" << *fullMove.moves->at(3);
//
//         return ostr;
//     }
//
//     ostr << "EMPTY FULLMOVE with score = " << fullMove.score;
//     return ostr;
// }
